---
title: "Identify DARS on each cell type"
---

*Description:*
This notebook will identify differentially accessible regions between CM, FB, EC, and MACRO cells. 
Once open regions on each cell type is identified, separate each cell type into individual objects. 
Within each cell type object, identify DARS between conditions. 
  - might have to split the object *further?*
  


```{r Load_libraries}
suppressPackageStartupMessages({
  library(Signac)
  library(Seurat)
  library(tidyverse)
  library(GenomicRanges)
  library(JASPAR2020)
  library(TFBSTools)
  library(BSgenome.Mmusculus.UCSC.mm10)
  library(clusterProfiler)
  library(org.Mm.eg.db)
  library(harmony)
  library(BiocParallel)
})

here::i_am("motif_analysis.Rmd")
set.seed(1)
```

## do not read this one
```{r read_object}
# preprocessed <- readRDS("/scratch/kwon.so/preprocessed_individual.rds")

DimPlot(preprocessed, reduction = "wnn.umap") + theme_classic()
```


# Adding motif information
```{r}
DefaultAssay(preprocessed) <- "ATAC"

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
preprocessed <- AddMotifs(
  object = preprocessed,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  pfm = pfm
)
```


```{r}
# saveRDS(preprocessed, file = "/scratch/kwon.so/motif_object.rds")
# preprocessed <- readRDS("/scratch/kwon.so/motif_object.rds")
```


# Link peaks to genes  
```{r eval=FALSE, include=FALSE}
DefaultAssay(preprocessed) <- "ATAC"

# computing GC content for each peak
preprocessed <- RegionStats(preprocessed, genome = BSgenome.Mmusculus.UCSC.mm10)

# link peaks to genes
# preprocessed <- LinkPeaks(
#   object = preprocessed,
#   peak.assay = "ATAC",
#   expression.assay = "RNA",
#   genes.use = NULL
# )

Links(preprocessed)
```


```{r}
# write_rds(preprocessed, "/scratch/kwon.so/linked_motif.rds")
preprocessed <- read_rds("/scratch/kwon.so/linked_motif.rds")
```



# Differential TF activity with Chromvar ---- directly finding diff motifs instead of finding DARS and then find motifs. 
# Subset and process cell types 
```{r}
register(MulticoreParam(8, progressbar = TRUE))
DefaultAssay(preprocessed) <- "ATAC"

# subset for individual cell types
cm_seu <- subset(preprocessed, subset = cell_type == "CM")
fb_seu <- subset(preprocessed, subset = cell_type == "FB")
ec_seu <- subset(preprocessed, subset = cell_type == "EC")
macro_seu <- subset(preprocessed, subset = cell_type == "Macro")

# process ATAC assay
process_atac <- function(seu_obj, cut_off = 20) {
  DefaultAssay(seu_obj) <- "ATAC"
  
  seu_obj <- RunTFIDF(seu_obj)
  seu_obj <- FindTopFeatures(seu_obj, min.cutoff = cut_off)
  seu_obj <- RunSVD(seu_obj)
  
  return(seu_obj)
}

cm_seu <- process_atac(cm_seu)
fb_seu <- process_atac(fb_seu)
ec_seu <- process_atac(ec_seu)
macro_seu <- process_atac(macro_seu)

Idents(cm_seu) <- cm_seu$condition
Idents(fb_seu) <- fb_seu$condition
Idents(ec_seu) <- ec_seu$condition
Idents(macro_seu) <- macro_seu$condition

cm_seu <- RegionStats(cm_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
fb_seu <- RegionStats(fb_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
ec_seu <- RegionStats(ec_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
macro_seu <- RegionStats(macro_seu, genome = BSgenome.Mmusculus.UCSC.mm10)

# running chromvar
cm_seu <- RunChromVAR(object = cm_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
fb_seu <- RunChromVAR(object = fb_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
ec_seu <- RunChromVAR(object = ec_seu, genome = BSgenome.Mmusculus.UCSC.mm10)
macro_seu <- RunChromVAR(object = macro_seu, genome = BSgenome.Mmusculus.UCSC.mm10)

###############################################################
# write_rds(cm_seu, "/scratch/kwon.so/chromvar_cm_seu.rds")
# write_rds(fb_seu, "/scratch/kwon.so/chromvar_fb_seu.rds")
# write_rds(ec_seu, "/scratch/kwon.so/chromvar_ec_seu.rds")
# write_rds(macro_seu, "/scratch/kwon.so/chromvar_macro_seu.rds")

cm_seu <- read_rds("/scratch/kwon.so/chromvar_cm_seu.rds")
fb_seu <- read_rds("/scratch/kwon.so/chromvar_fb_seu.rds")
ec_seu <- read_rds("/scratch/kwon.so/chromvar_ec_seu.rds")
macro_seu <- read_rds("/scratch/kwon.so/chromvar_macro_seu.rds")
###############################################################


DefaultAssay(cm_seu) <- "chromvar"
DefaultAssay(fb_seu) <- "chromvar"
DefaultAssay(ec_seu) <- "chromvar"
DefaultAssay(macro_seu) <- "chromvar"

# map motifs to motif names
motif_names <- GetMotifData(preprocessed[["ATAC"]], slot = "motif.names") %>%
  data.frame() %>% 
  pivot_longer(cols = everything(), names_to = "motif", values_to = "motif.name")

## CM
motifs_cm_t <- FindMarkers(cm_seu,
                             ident.1 = "TAC",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

motifs_cm_a <- FindMarkers(cm_seu,
                          ident.1 = "28M",
                          ident.2 = "4M-H",
                          test.use = "wilcox",
                          mean.fxn = rowMeans,
                          fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

## FB
motifs_fb_t <- FindMarkers(fb_seu,
                             ident.1 = "TAC",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

motifs_fb_a <- FindMarkers(fb_seu,
                          ident.1 = "28M",
                          ident.2 = "4M-H",
                          test.use = "wilcox",
                          mean.fxn = rowMeans,
                          fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

## EC
motifs_ec_t <- FindMarkers(ec_seu,
                             ident.1 = "TAC",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

motifs_ec_a <- FindMarkers(ec_seu,
                          ident.1 = "28M",
                          ident.2 = "4M-H",
                          test.use = "wilcox",
                          mean.fxn = rowMeans,
                          fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

## macro
motifs_macro_t <- FindMarkers(macro_seu,
                             ident.1 = "TAC",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

motifs_macro_a <- FindMarkers(macro_seu,
                          ident.1 = "28M",
                          ident.2 = "4M-H",
                          test.use = "wilcox",
                          mean.fxn = rowMeans,
                          fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)


# selecting significant motifs
sig_motifs_cm_t <- motifs_cm_t %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)
sig_motifs_cm_a <- motifs_cm_a %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)

sig_motifs_fb_t <- motifs_fb_t %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)
sig_motifs_fb_a <- motifs_fb_a %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)

sig_motifs_ec_t <- motifs_ec_t %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)
sig_motifs_ec_a <- motifs_ec_a %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)

sig_motifs_macro_t <- motifs_macro_t %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)
sig_motifs_macro_a <- motifs_macro_a %>% dplyr::filter(p_val_adj < 0.05, avg_diff > 0)



# find unique and shared motifs between conditions
unique_cm_t <- setdiff(sig_motifs_cm_t$motif.name, sig_motifs_cm_a$motif.name)
unique_cm_a <- setdiff(sig_motifs_cm_a$motif.name, sig_motifs_cm_t$motif.name) 
shared_cm <- intersect(sig_motifs_cm_t$motif.name, sig_motifs_cm_a$motif.name) 

unique_fb_t <- setdiff(sig_motifs_fb_t$motif.name, sig_motifs_fb_a$motif.name) 
unique_fb_a <- setdiff(sig_motifs_fb_a$motif.name, sig_motifs_fb_t$motif.name) 
shared_fb <- intersect(sig_motifs_fb_t$motif.name, sig_motifs_fb_a$motif.name) 

unique_ec_t <- setdiff(sig_motifs_ec_t$motif.name, sig_motifs_ec_a$motif.name) 
unique_ec_a <- setdiff(sig_motifs_ec_a$motif.name, sig_motifs_ec_t$motif.name) 
shared_ec <- intersect(sig_motifs_ec_t$motif.name, sig_motifs_ec_a$motif.name) 

unique_macro_t <- setdiff(sig_motifs_macro_t$motif.name, sig_motifs_macro_a$motif.name)
unique_macro_a <- setdiff(sig_motifs_macro_a$motif.name, sig_motifs_macro_t$motif.name)
shared_macro <- intersect(sig_motifs_macro_t$motif.name, sig_motifs_macro_a$motif.name)

unique_list <- list(cm_t = unique_cm_t, cm_a = unique_cm_a, cm_shared = shared_cm,
                    fb_t = unique_fb_t, fb_a = unique_fb_a, fb_shared = shared_fb,
                    ec_t = unique_ec_t, ec_a = unique_ec_a, ec_shared = shared_ec,
                    macro_t = unique_macro_t, macro_a = unique_macro_a, macro_shared = shared_macro)


##### selecting motif names
cmt <- sig_motifs_cm_t %>% dplyr::filter(motif.name %in% unique_cm_t)
cma <- sig_motifs_cm_a %>% dplyr::filter(motif.name %in% unique_cm_a)
cms <- sig_motifs_cm_t %>% dplyr::filter(motif.name %in% shared_cm)

fbt <- sig_motifs_fb_t %>% dplyr::filter(motif.name %in% unique_fb_t)
fba <- sig_motifs_fb_a %>% dplyr::filter(motif.name %in% unique_fb_a)
fbs <- sig_motifs_fb_t %>% dplyr::filter(motif.name %in% shared_fb)

ect <- sig_motifs_ec_t %>% dplyr::filter(motif.name %in% unique_ec_t)
eca <- sig_motifs_ec_a %>% dplyr::filter(motif.name %in% unique_ec_a)
ecs <- sig_motifs_ec_t %>% dplyr::filter(motif.name %in% shared_ec)

mat <- sig_motifs_macro_t %>% dplyr::filter(motif.name %in% unique_macro_t)
maa <- sig_motifs_macro_a %>% dplyr::filter(motif.name %in% unique_macro_a)
mas <- sig_motifs_macro_t %>% dplyr::filter(motif.name %in% shared_macro)


# intersecting motifs conditions
# cm_venn <- ggVennDiagram::ggVennDiagram(list(sig_motifs_cm_t$motif.name, sig_motifs_cm_a$motif.name),
#                              category.names = c("TAC", "Aged")) + 
#   labs(title = "CM Enriched Motifs between condition")
# 
# fb_venn <- ggVennDiagram::ggVennDiagram(list(sig_motifs_fb_t$motif.name, sig_motifs_fb_a$motif.name),
#                              category.names = c("TAC", "Aged")) + 
#   labs(title = "FB Enriched Motifs between condition")
# 
# ec_venn <- ggVennDiagram::ggVennDiagram(list(sig_motifs_ec_t$motif.name, sig_motifs_ec_a$motif.name),
#                              category.names = c("TAC", "Aged")) + 
#   labs(title = "EC Enriched Motifs between condition")
# 
# macro_venn <- ggVennDiagram::ggVennDiagram(list(sig_motifs_macro_t$motif.name, sig_motifs_macro_a$motif.name),
#                              category.names = c("TAC", "Aged")) + 
#   labs(title = "Macro Enriched Motifs between condition")



####### download motifs to csv and make a venn diagram and see what the differences are 
# TAC espefici and AGEd specific 
#### if there is any motif in TAC that is in fibrosys 
### look at functions of TFs. 
# if there are TF shared between cell types then run module score of those TFs 
# show expression of the shared one to show if there is contaminaiton or if the expression is high in cM then the reguation is higher in cms



setdiff(unique_cm_t, c(unique_fb_t, unique_ec_t, unique_macro_t))  # unique motifs for CM TAC
setdiff(unique_cm_a, c(unique_fb_a, unique_ec_a, unique_macro_a))  # unique motifs CM aged

setdiff(unique_cm_t, unique_ec_t) # unique motifs in CM TAC v. EC
setdiff(unique_cm_t, unique_macro_t) # CM TAC v. macro 
```


## CM
```{r}
DefaultAssay(cm_seu) <- "ATAC"

#### LR test
# cm_lr_TvH <- FindMarkers(
#   object = cm_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# write_rds(cm_lr_TvH, "/scratch/kwon.so/cm_lr_TvH.rds")
# 
# cm_lr_AvH <- FindMarkers(
#   object = cm_seu,
#   ident.1 = '28M',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# write_rds(cm_lr_AvH, "/scratch/kwon.so/cm_lr_AvH.rds")

cm_lr_TvH <- read_rds("/scratch/kwon.so/cm_lr_TvH.rds")
cm_lr_AvH <- read_rds("/scratch/kwon.so/cm_lr_AvH.rds")
  

# top differential accessible regions 
top_cm_TvH <- cm_lr_TvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)
top_cm_AvH <- cm_lr_AvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)


# background matching peaks
open_peaks_tac <- AccessiblePeaks(cm_seu, idents = c("TAC", "4M-H"), assay = "ATAC")
open_peaks_aged <- AccessiblePeaks(cm_seu, idents = c("28M", "4M-H"), assay = "ATAC")

# match the overall GC content in the peak set
meta.feature <- GetAssayData(cm_seu, assay = "ATAC", layer = "meta.features")

peaks_matched_tac <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_tac, ],
  query.feature = meta.feature[rownames(top_cm_TvH), ],
  n = 50000
)

peaks_matched_aged <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_aged, ],
  query.feature = meta.feature[rownames(top_cm_AvH), ],
  n = 50000
)

# test enrichment
cm_TvH_motif <- FindMotifs(
  object = cm_seu,
  features = rownames(top_cm_TvH),
  background = peaks_matched_tac,
  assay = "ATAC"
)

cm_AvH_motif <- FindMotifs(
  object = cm_seu,
  features = rownames(top_cm_AvH),
  background = peaks_matched_aged,
  assay = "ATAC"
)

# filtering for significant motifs 
cm_TvH_motif %>% dplyr::filter(p.adjust < 0.05) 
cm_AvH_motif %>% dplyr::filter(p.adjust < 0.05)

# compare motifs between LR and chromvar ---------------------------- use the shared ones to identify where they bind to 
cm_lr_t_motifs <- cm_TvH_motif %>% dplyr::filter(p.adjust < 0.05) 
cm_lr_a_motifs <- cm_AvH_motif %>% dplyr::filter(p.adjust < 0.05)

cm_t_shared_motif <- inner_join(cm_lr_t_motifs, sig_motifs_cm_t, by = join_by(motif.name))
cm_a_shared_motif <- inner_join(cm_lr_a_motifs, sig_motifs_cm_a, by = join_by(motif.name))
```



### plot of top motifs
```{r}
filtered_motifs <- cm_TvH_motif %>% dplyr::filter(p.adjust < 0.05)

ggplot(filtered_motifs, aes(x = reorder(motif.name, fold.enrichment), y = fold.enrichment)) +
    geom_bar(stat = "identity") +
    # coord_flip() +
    theme_minimal() +
    labs(title = paste("Top Enriched Motifs in CMs"),
         x = "Motif",
         y = "Fold Enrichment") + theme(axis.text.x = element_text(angle = 90)) 


###### check that the upregulated DEGs in scRNA have enriched motifs here. 
# could show that TFs are active here. 
cm_TvH_motif %>% dplyr::filter(motif.name %in% c("DMD", "MYH7", "ANKRD1", "CCN2", "FGFR2", "CISH", "CECPB", "SLC8A1"))


##### scenic
# SCENIC analysis setup
dbFiles <- c('https://resources.aertslab.org/cistarget/databases/old/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather', 'https://resources.aertslab.org/cistarget/databases/old/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-tss-centered-5kb-7species.mc9nr.feather')

for (file in dbFiles) {
  download.file(file, destfile = basename(file))
}

exprMat <- as.matrix(cm_seu@assays$RNA$counts)
cellInfo <- data.frame(cm_seu@meta.data)
scenicOptions <- initializeScenic(org = "mgi", dbDir = "cisTarget_databases", nCores = 8, 
                                  dbs = c("mm9-500bp-upstream-7species.mc9nr.feather",
                                          "mm9-tss-centered-5kb-7species.mc9nr.feather"))

# gene filtering
genesKept <- geneFiltering(exprMat, scenicOptions=scenicOptions,
                           minCountsPerGene=3*.01*ncol(exprMat),
                           minSamples=ncol(exprMat)*.01)


exprMat_filtered <- exprMat[genesKept, ]
exprMat_log <- log2(exprMat_filtered + 1)

# run scenic pipeline
# Co-expression network
runGenie3(exprMat_log, scenicOptions)

# Module construction
runSCENIC_1_coexNetwork2modules(scenicOptions)

# TF motif enrichment + Regulon definition
runSCENIC_2_createRegulons(scenicOptions)

# Score regulons per cell
runSCENIC_3_scoreCells(scenicOptions, exprMat_log)

# add scenic actitivty to seurat
# Load AUC matrix
regulonAUC <- loadInt(scenicOptions, "aucell_regulonAUC")
aucMat <- getAUC(regulonAUC)  # regulons x cells

# Make sure column names match
aucMat <- aucMat[, colnames(seu)]



```


### pathway enrichment for CM
Performing pathway enrichment analysis on the upregulated regions of TAC samples and Aged samples. 

For TAC samples, enriched pathways: Wnt signaling, 
Aged samples enriched pathways: Cardiac muscle hypertrophy, muscle cell differentiation, 
```{r}
# closest feature for significant CM peaks
open_cm_TvH <- ClosestFeature(cm_seu, rownames(top_cm_TvH))
open_cm_AvH <- ClosestFeature(cm_seu, rownames(top_cm_AvH))

# pathway enrichment
cm_TvH_ego <- enrichGO(gene = open_cm_TvH$gene_name,
                keyType = "SYMBOL",
                OrgDb = org.Mm.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)

barplot(cm_TvH_ego, showCategory = 20)


cm_AvH_ego <- enrichGO(gene = open_cm_AvH$gene_name,
                keyType = "SYMBOL",
                OrgDb = org.Mm.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)

barplot(cm_AvH_ego, showCategory = 20)
```



## FB 
```{r}
DefaultAssay(fb_seu) <- "ATAC"

### LR test
# fb_lr_TvH <- FindMarkers(
#   object = fb_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# fb_lr_AvH <- FindMarkers(
#   object = fb_seu,
#   ident.1 = '28M',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# write_rds(fb_lr_TvH, "/scratch/kwon.so/fb_lr_TvH.rds")
# write_rds(fb_lr_AvH, "/scratch/kwon.so/fb_lr_AvH.rds")

fb_lr_TvH <- read_rds("/scratch/kwon.so/fb_lr_TvH.rds")
fb_lr_AvH <- read_rds("/scratch/kwon.so/fb_lr_AvH.rds")

# top differentially accessible peaks 
top_fb_TvH <- fb_lr_TvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)
top_fb_AvH <- fb_lr_AvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)


# background match peaks
open_peaks_tac <- AccessiblePeaks(fb_seu, idents = c("TAC", "4M-H"), assay = "ATAC")
open_peaks_aged <- AccessiblePeaks(fb_seu, idents = c("28M", "4M-H"), assay = "ATAC")

# match the overall GC content in the peak set
meta.feature <- GetAssayData(fb_seu, assay = "ATAC", layer = "meta.features")

peaks_matched_tac <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_tac, ],
  query.feature = meta.feature[rownames(top_fb_TvH), ],
  n = 50000
)

peaks_matched_aged <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_aged, ],
  query.feature = meta.feature[rownames(top_fb_AvH), ],
  n = 50000
)


# enriched motifs
fb_TvH_motif <- FindMotifs(object = fb_seu,
                           features = rownames(top_fb_TvH),
                          background = peaks_matched_tac, assay = "ATAC")

fb_AvH_motif <- FindMotifs(object = fb_seu,
                           features = rownames(top_fb_AvH),
                           background = peaks_matched_aged, assay = "ATAC")


# significant motifs
fb_TvH_motif %>% dplyr::filter(pvalue < 0.05)
fb_AvH_motif %>% dplyr::filter(pvalue < 0.05)

# compare motifs between LR and chromvar

```

### FB pathway enrichment
```{r}
open_fb_TvH <- ClosestFeature(fb_seu, rownames(top_fb_TvH))
open_fb_AvH <- ClosestFeature(fb_seu, rownames(top_fb_AvH))

fb_TvH_ego <- enrichGO(gene = open_fb_TvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(fb_TvH_ego,showCategory = 20)


fb_AvH_ego <- enrichGO(gene = open_fb_AvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(fb_AvH_ego, showCategory = 20)
```


## EC 
```{r}
DefaultAssay(ec_seu) <- "ATAC"

### LR test
# ec_lr_TvH <- FindMarkers(
#   object = ec_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# ec_lr_AvH <- FindMarkers(
#   object = ec_seu,
#   ident.1 = '28M',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# write_rds(ec_lr_TvH, "/scratch/kwon.so/ec_lr_TvH.rds")
# write_rds(ec_lr_AvH, "/scratch/kwon.so/ec_lr_AvH.rds")

ec_lr_TvH <- read_rds("/scratch/kwon.so/ec_lr_TvH.rds")
ec_lr_AvH <- read_rds("/scratch/kwon.so/ec_lr_AvH.rds")


# top diff accessible peaks
top_ec_TvH <- ec_lr_TvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)
top_ec_AvH <- ec_lr_AvH[ec_lr_AvH$p_val < 0.005 & ec_lr_AvH$avg_log2FC > 0.25, ]


# background match peaks
open_peaks_tac <- AccessiblePeaks(ec_seu, idents = c("TAC", "4M-H"), assay = "ATAC")
open_peaks_aged <- AccessiblePeaks(ec_seu, idents = c("28M", "4M-H"), assay = "ATAC")

# match the overall GC content in the peak set
meta.feature <- GetAssayData(ec_seu, assay = "ATAC", layer = "meta.features")

peaks_matched_tac <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_tac, ],
  query.feature = meta.feature[rownames(top_ec_TvH), ],
  n = 50000
)

peaks_matched_aged <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_aged, ],
  query.feature = meta.feature[rownames(top_ec_AvH), ],
  n = 50000
)


# enriched motifs
ec_TvH_motif <- FindMotifs(ec_seu,
                          features = rownames(top_ec_TvH),
                          background = peaks_matched_tac,
                          assay = "ATAC")

ec_AvH_motif <- FindMotifs(ec_seu,
                          features = rownames(top_ec_AvH),
                          background = peaks_matched_aged,
                          assay = "ATAC")

# filter for significant motifs
ec_TvH_motif %>% dplyr::filter(p.adjust < 0.05)
ec_AvH_motif %>% dplyr::filter(p.adjust < 0.05)

# compare motifs between LR and chromvar

```

### EC pathway enrichment 
```{r}
open_ec_TvH <- ClosestFeature(ec_seu, rownames(top_ec_TvH))
open_ec_AvH <- ClosestFeature(ec_seu, rownames(top_ec_AvH))

ec_TvH_ego <- enrichGO(gene = open_ec_TvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(ec_TvH_ego,showCategory = 20)


ec_AvH_ego <- enrichGO(gene = open_ec_AvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(ec_AvH_ego, showCategory = 20)
```


##### Next step would be to identify where this motifs bind to. 
# Motif footprinting
```{r}
Footprint(object = preprocessed, motif.name = , genome = BSgenome.Mmusculus.UCSC.mm10)
```


# TF footprinting with motifs ----- useless i think......... just takes a lot to run and only shows a plot. its more for individual genes i want to check 
Analyzes sequences to identify regions where TFs bind to DNA. 
```{r}
DefaultAssay(cm_seu) <- "ATAC"

# gather the footprinting information for sets of motifs
test <- Footprint(
  object = cm_seu,
  motif.name = motifs_cm_t$motif.name[1:3],
  genome = BSgenome.Mmusculus.UCSC.mm10
)

# plot the footprint data for each group of cells
PlotFootprint(test, features = motifs_cm_t$motif.name[1:3])
```




```{r eval=FALSE, include=FALSE}
# save csv of significant enriched motifs
# write_csv(sig_cm_4T_4H, file = "./motif_results/CM_4Tv4H.csv")
# write_csv(sig_cm_28M_4H, file = "./motif_results/CM_28Mv4H.csv")
# 
# write_csv(sig_fb_4T, file = "./motif_results/FB_4H_4T.csv")
# write_csv(sig_fb_28M, file = "./motif_results/FB_4H_28M.csv")
# 
# write_csv(sig_ec_4Hvs4T, file = "./motif_results/EC_4H_4T.csv")
# write_csv(sig_ec_4Hvs28M, file = "./motif_results/EC_4H_28M.csv")
```


