---
title: "Identify DARS on each cell type"
---

*Description:*
This notebook will identify differentially accessible regions between CM, FB, EC, and MACRO cells. 
Once open regions on each cell type is identified, separate each cell type into individual objects. 
Within each cell type object, identify DARS between conditions. 
  - might have to split the object *further?*
  
* Not sure if I need to identify DARS between the cell types (1st step). 


```{r Load_libraries}
suppressPackageStartupMessages({
  library(Signac)
  library(Seurat)
  library(tidyverse)
  library(GenomicRanges)
  library(JASPAR2020)
  library(TFBSTools)
  library(BSgenome.Mmusculus.UCSC.mm10)
  library(clusterProfiler)
  library(org.Mm.eg.db)
  library(harmony)
  library(BiocParallel)
})

here::i_am("motif_analysis.Rmd")
set.seed(1)
```


```{r read_object}
preprocessed <- readRDS("/scratch/kwon.so/preprocessed_individual.rds")

DimPlot(preprocessed, reduction = "wnn.umap") + theme_classic()
```


# Adding motif information
```{r}
DefaultAssay(preprocessed) <- "ATAC"

# Get a list of motif position frequency matrices from the JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

# add motif information
preprocessed <- AddMotifs(
  object = preprocessed,
  genome = BSgenome.Mmusculus.UCSC.mm10,
  pfm = pfm
)
```


```{r}
# saveRDS(preprocessed, file = "/scratch/kwon.so/motif_object.rds")
preprocessed <- readRDS("/scratch/kwon.so/motif_object.rds")
```


# Link peaks to genes --- takes a long time to run with whole genome 
```{r}
DefaultAssay(preprocessed) <- "ATAC"

# computing GC content for each peak
preprocessed <- RegionStats(preprocessed, genome = BSgenome.Mmusculus.UCSC.mm10)

# link peaks to genes
preprocessed <- LinkPeaks(
  object = preprocessed,
  peak.assay = "ATAC",
  expression.assay = "RNA",
  genes.use = c("Nfe2l1")
)
```


# Subset cell types
Need to process the data again after subset from the complete object. 
Harmony can be run again after subsetting and will not change the RNA assay. 

      Data is not wrong but there is not enough information to determine whether data was integrated. 
      If we had 24M, we would be able to see that in between 4M and 28M which would tell us that its done correctly. 
```{r}
cm <- subset(preprocessed, subset = cell_type == "CM")

DefaultAssay(cm) <- "RNA"

cm <- NormalizeData(cm)
cm <- FindVariableFeatures(cm)
cm <- ScaleData(cm)
cm <- RunPCA(cm)
cm <- FindNeighbors(cm, dims = 1:20, reduction = "pca")
cm <- FindClusters(cm, resolution = 0.5)
cm <- RunUMAP(cm, dims = 1:30, reduction = "pca", reduction.name = "cm.umap")
DimPlot(cm, reduction = "cm.umap", group.by = "condition", label = T)

# run harmony again since the dataset is not integrated
cm <- RunHarmony(object = cm, group.by.vars = "dataset")
cm <- RunUMAP(cm, dims = 1:30, reduction = "harmony", reduction.name = "harmony.UMAP")
DimPlot(cm, reduction = "harmony.UMAP", label = T, split.by = "condition")



```




# Differential regions: Cell types
CM vs FB; CM vs EC; CM vs MACRO
```{r}
cm_peaks <- FindMarkers(object = preprocessed,
                        ident.1 = "CM",
                        ident.2 = c("FB", "EC", "Macro"),
                        test.use = 'LR',
                        min.pct = 0.05,
                        latent.vars = "nCount_ATAC")

fb_peaks <- FindMarkers(preprocessed, 
                        ident.1 = "FB", 
                        ident.2 = c("CM", "EC", "Macro"), 
                        test.use = "LR", 
                        min.pct = 0.05,
                        latent.vars = "nCount_ATAC")

ec_peaks <- FindMarkers(preprocessed,
                        ident.1 = "EC", 
                        ident.2 = c("CM", "FB", "Macro"),
                        test.use = "LR", 
                        min.pct = 0.05,
                        latent.vars = "nCount_ATAC")

macro_peaks <- FindMarkers(preprocessed,
                           ident.1 = "Macro",
                           ident.2 = c("CM", "FB", "EC"),
                           test.use = "LR",
                           min.pct = 0.05,
                           latent.vars = "nCount_ATAC")

top_da_cm <- cm_peaks[cm_peaks$p_val_adj < 0.05 & cm_peaks$avg_log2FC > 3, ]
top_da_fb <- fb_peaks[fb_peaks$p_val_adj < 0.05 & fb_peaks$avg_log2FC > 1, ]
top_da_ec <- ec_peaks[ec_peaks$p_val_adj < 0.05 & ec_peaks$avg_log2FC > 1, ]
top_da_macro <- macro_peaks[macro_peaks$p_val_adj < 0.05 & macro_peaks$avg_log2FC > 1, ]

cm_motif <- FindMotifs(preprocessed, features = rownames(top_da_cm))
fb_motif <- FindMotifs(preprocessed, features = rownames(top_da_fb))
ec_motif <- FindMotifs(preprocessed, features = rownames(top_da_ec))
macro_motif <- FindMotifs(preprocessed, features = rownames(top_da_macro))

# filter motifs
cm_motif %>% dplyr::filter(fold.enrichment > 1.25 & p.adjust < 0.05)
fb_motif %>% dplyr::filter(fold.enrichment > 1.5 & p.adjust < 0.05)
ec_motif %>% dplyr::filter(fold.enrichment > 1.5 & p.adjust < 0.05)
macro_motif %>% dplyr::filter(fold.enrichment > 1.5 & p.adjust < 0.05)
```


# Differential regions: Conditions --- Not really ideal. Focus more on individual cell types. 
```{r}
Idents(preprocessed) <- preprocessed$condition

# TACvHealth_peak <- FindMarkers(object = preprocessed, 
#                                 ident.1 = "TAC",
#                                 ident.2 = "4M-H",
#                                 test.use = 'LR',
#                                 min.pct = 0.05,
#                                latent.vars = "nCount_ATAC")

# saveRDS(TACvHealth_peak, file = "/scratch/kwon.so/TACvHealth_peak.rds")
TACvHealth_peak <- readRDS("/scratch/kwon.so/TACvHealth_peak.rds")

# AgedvHealth_peak <- FindMarkers(object = preprocessed,
#                                 ident.1 = "28M",
#                                 ident.2 = "4M-H",
#                                 test.use = 'LR',
#                                 min.pct = 0.05,
#                                 latent.vars = "nCount_ATAC")

# saveRDS(AgedvHealth_peak, file = "/scratch/kwon.so/AgedvHealth_peak.rds")
AgedvHealth_peak <- readRDS("/scratch/kwon.so/AgedvHealth_peak.rds")


# getting top differentially accessible peaks 
top_TvH <- TACvHealth_peak %>% dplyr::filter(p_val_adj < 0.05, avg_log2FC > 1)
top_AvH <- AgedvHealth_peak %>% dplyr::filter(p_val_adj < 0.05, avg_log2FC > 1)

# finding enriched motifs
# this only compares the peaks input to matched background peaks
TvH_motif <- FindMotifs(object = preprocessed,
                        features = rownames(top_TvH))

AvH_motif <- FindMotifs(object = preprocessed,
                        features = rownames(top_AvH))

####### need to run footprinting --- check any enriched motifs using the footprinting functions 
# identifies regions where TFs are likely bound to DNA



## Filter enriched motifs
TvH_motif %>% dplyr::filter(fold.enrichment > 1.5, p.adjust < 0.05)
AvH_motif %>% dplyr::filter(fold.enrichment > 1.5, p.adjust < 0.05)

# 
tac_unique <- setdiff(rownames(top_TvH), rownames(top_AvH)) # unique in TAC
aged_unique <- setdiff(rownames(top_AvH), rownames(top_TvH)) # unique in aged
shared_tac_aged <- intersect(rownames(top_TvH), rownames(top_AvH)) # shared regions


##### just for show
VlnPlot(
  object = preprocessed,
  features = rownames(top_TvH)[1],
  pt.size = 0.001,
  idents = c("TAC", "28M", "4M-H")
)


```

### Pathway enrichment
```{r}
# rn choosing the upregulated accessible regions
open_T <- ClosestFeature(preprocessed, rownames(top_TvH))
open_A <- ClosestFeature(preprocessed, rownames(top_AvH))


tac_h_ego <- enrichGO(gene = open_T$gene_name,
                      keyType = "SYMBOL",
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)

age_h_ego <- enrichGO(gene = open_A$gene_name,
                      keyType = "SYMBOL",
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)

barplot(tac_h_ego, showCategory = 20)
barplot(age_h_ego, showCategory = 20)


##### overexpressed pathways on each condition and shared
tac_close_genes <- ClosestFeature(preprocessed, regions = tac_unique)
aged_close_genes <- ClosestFeature(preprocessed, regions = aged_unique)
shared_close_genes <- ClosestFeature(preprocessed, regions = shared_tac_aged)

tac_ego <- enrichGO(gene = tac_close_genes$gene_name,
                    keyType = "SYMBOL",
                    OrgDb = org.Mm.eg.db,
                    ont = "BP",
                    qvalueCutoff = 0.05,
                    readable = TRUE)

aged_ego <- enrichGO(gene = aged_close_genes$gene_name,
                    keyType = "SYMBOL",
                    OrgDb = org.Mm.eg.db,
                    ont = "BP",
                    qvalueCutoff = 0.05,
                    readable = TRUE)

shared_ego <- enrichGO(gene = shared_close_genes$gene_name,
                    keyType = "SYMBOL",
                    OrgDb = org.Mm.eg.db,
                    ont = "BP",
                    qvalueCutoff = 0.05,
                    readable = TRUE)

barplot(tac_ego, showCategory = 20)
barplot(aged_ego, showCategory = 20)
barplot(shared_ego, showCategory = 20)

dotplot(tac_ego) + ggtitle("M4T overrepresentation")
dotplot(aged_ego) + ggtitle("M28 sample pathways")

```
When doing ORA on both upregulated and downregulated regions --> So it seems like aged samples have an overrepresentation of metabolic processes such as phosphorus metabolic processes, and phosphorylation mechanisms.

Performed overrepresentation analysis on upregulated regions in TAC only, aged only, and shared regions. 
TAC only regions seem to have an over representation of 


# Differential TF activity with Chromvar 
```{r}
register(MulticoreParam(8, progressbar = TRUE))

# preprocessed <- RunChromVAR(
#   object = preprocessed,
#   genome = BSgenome.Mmusculus.UCSC.mm10
# )

# saveRDS(preprocessed, file = "/scratch/kwon.so/preprocessed_chromvar.rds")
preprocessed <- readRDS("/scratch/kwon.so/preprocessed_chromvar.rds")

DefaultAssay(preprocessed) <- "chromvar"

# map motifs to motif names
motif_names <- GetMotifData(preprocessed[["ATAC"]], slot = "motif.names") %>%
  data.frame() %>% 
  pivot_longer(cols = everything(), names_to = "motif", values_to = "motif.name")

motif_markers_tac <- FindMarkers(preprocessed,
                             ident.1 = "TAC",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)

motif_markers_age <- FindMarkers(preprocessed,
                             ident.1 = "28M",
                             ident.2 = "4M-H",
                             test.use = "wilcox",
                             mean.fxn = rowMeans,
                             fc.name = "avg_diff") %>%
  rownames_to_column("motif") %>% 
  left_join(motif_names)



motif_markers_tac %>% arrange(avg_diff)
motif_markers_age

MotifPlot(preprocessed, motifs = head(motif_markers_tac$motif.name))
MotifPlot(preprocessed, motifs = head(motif_markers_age$motif.name))

DefaultAssay(preprocessed) <- "ATAC"

CoveragePlot(preprocessed, region = "Etv2")
```



# Differential regions: Conditions on each cell type
```{r}
Idents(preprocessed) <- preprocessed$condition
```

## CM  

Try background matching for peaks when running FindMarkers. 

```{r}
cm_seu <- subset(preprocessed, subset = cell_type == "CM")

#### LR test

# cm_lr_TvH <- FindMarkers( 
#   object = cm_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# saveRDS(cm_lr_TvH, "/scratch/kwon.so/cm_lr_TvH.rds")
cm_lr_TvH <- readRDS("/scratch/kwon.so/cm_lr_TvH.rds")

# cm_lr_AvH <- FindMarkers
#   object = cm_seu,
#   ident.1 = '28M',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# saveRDS(cm_lr_AvH, "/scratch/kwon.so/cm_lr_AvH.rds")
cm_lr_AvH <- readRDS("/scratch/kwon.so/cm_lr_AvH.rds")
  

# top differential accessible regions
top_cm_TvH <- cm_lr_TvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 1)
top_cm_AvH <- cm_lr_AvH %>% dplyr::filter(p_val < 0.005, avg_log2FC > 0.25)


# background matching peaks
open_peaks_tac <- AccessiblePeaks(cm_seu, idents = c("TAC", "4M-H"))
open_peaks_aged <- AccessiblePeaks(cm_seu, idents = c("28M", "4M-H"))

# match the overall GC content in the peak set
meta.feature <- GetAssayData(cm_seu, assay = "ATAC", layer = "meta.features")

peaks_matched_tac <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_tac, ],
  query.feature = meta.feature[rownames(top_cm_TvH), ],
  n = 50000
)

peaks_matched_aged <- MatchRegionStats(
  meta.feature = meta.feature[open_peaks_aged, ],
  query.feature = meta.feature[rownames(top_cm_AvH), ],
  n = 50000
)

# test enrichment
cm_TvH_motif <- FindMotifs(
  object = cm_seu,
  features = rownames(top_cm_TvH),
  background = peaks_matched_tac
)

cm_AvH_motif <- FindMotifs(
  object = cm_seu,
  features = rownames(top_cm_AvH),
  background = peaks_matched_aged
)

# filtering for significant and enriched motifs ---------- Don't think this does anything...
# cm_TvH_motif %>% dplyr::filter(fold.enrichment > 1.6 & p.adjust < 0.05)
# cm_AvH_motif %>% dplyr::filter(fold.enrichment > 1.5 & p.adjust < 0.05)
```



### pathway enrichment for CM
Performing pathway enrichment analysis on the upregulated regions of TAC samples and Aged samples. 

For TAC samples, enriched pathways: Wnt signaling, 
Aged samples enriched pathways: Cardiac muscle hypertrophy, muscle cell differentiation, 
```{r}
# closest feature for significant CM peaks
open_cm_TvH <- ClosestFeature(cm_seu, rownames(top_cm_TvH))
open_cm_AvH <- ClosestFeature(cm_seu, rownames(top_cm_AvH))

# pathway enrichment
cm_TvH_ego <- enrichGO(gene = open_cm_TvH$gene_name,
                keyType = "SYMBOL",
                OrgDb = org.Mm.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)

barplot(cm_TvH_ego, showCategory = 20)


cm_AvH_ego <- enrichGO(gene = open_cm_AvH$gene_name,
                keyType = "SYMBOL",
                OrgDb = org.Mm.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                readable = TRUE)

barplot(cm_AvH_ego, showCategory = 20)


summary(cm_TvH_ego)
```



## FB
```{r}
fb_seu <- subset(preprocessed, subset = cell_type == "FB")

fb_TvH <- FindMarkers(object = fb_seu,
                      ident.1 = "TAC",
                      ident.2 = "4M-H",
                      test.use = 'wilcox',
                      min.pct = 0.1)

fb_AvH <- FindMarkers(object = fb_seu,
                      ident.1 = "28M",
                      ident.2 = "4M-H",
                      test.use = 'wilcox',
                      min.pct = 0.1)

### LR test
# fb_lr_TvH <- FindMarkers( 
#   object = fb_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# fb_lr_AvH <- FindMarkers( 
#   object = fb_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# saveRDS(fb_lr_TvH, "/scratch/kwon.so/fb_lr_TvH.rds")
# saveRDS(fb_lr_AvH, "/scratch/kwon.so/fb_lr_AvH.rds")

fb_lr_TvH <- readRDS("/scratch/kwon.so/fb_lr_TvH.rds")
fb_lr_AvH <- readRDS("/scratch/kwon.so/fb_lr_AvH.rds")

# top differentially accessible peaks
top_fb_TvH <- fb_lr_TvH[fb_lr_TvH$p_val_adj < 0.05 & fb_lr_TvH$avg_log2FC > 1.5, ]
top_fb_AvH <- fb_lr_AvH[fb_lr_AvH$p_val_adj < 0.05 & fb_lr_AvH$avg_log2FC > 0.5, ]


# enriched motifs
fb_TvH_motif <- FindMotifs(object = fb_seu,
                           features = rownames(top_fb_TvH))

fb_AvH_motif <- FindMotifs(object = fb_seu,
                           features = rownames(top_fb_AvH))


# filtering for fold.enrichment and p.adjust

```

### FB pathway enrichment
```{r}
open_fb_TvH <- ClosestFeature(fb_seu, rownames(top_fb_TvH))
open_fb_AvH <- ClosestFeature(fb_seu, rownames(top_fb_AvH))

fb_TvH_ego <- enrichGO(gene = open_fb_TvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(fb_TvH_ego,showCategory = 20)


fb_AvH_ego <- enrichGO(gene = open_fb_AvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(fb_AvH_ego, showCategory = 20)
```


## EC
```{r}
ec_seu <- subset(preprocessed, subset = cell_type == "EC")

# ec_TvH <- FindMarkers(
#   ec_seu,
#   ident.1 = "TAC",
#   ident.2 = "4M-H",
#   test.use = "wilcox",
#   min.pct = 0.05
# )
# 
# ec_AvH <- FindMarkers(
#   ec_seu,
#   ident.1 = "28M",
#   ident.2 = "4M-H",
#   test.use = "wilcox",
#   min.pct = 0.05
# )

### LR test
# ec_lr_TvH <- FindMarkers( 
#   object = ec_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# ec_lr_AvH <- FindMarkers( 
#   object = ec_seu,
#   ident.1 = 'TAC',
#   ident.2 = '4M-H',
#   test.use = 'LR',
#   min.pct = 0.05,
#   latent.vars = 'nCount_ATAC'
# )
# 
# saveRDS(ec_lr_TvH, "/scratch/kwon.so/ec_lr_TvH.rds")
# saveRDS(ec_lr_AvH, "/scratch/kwon.so/ec_lr_AvH.rds")

ec_lr_TvH <- readRDS("/scratch/kwon.so/ec_lr_TvH.rds")
ec_lr_AvH <- readRDS("/scratch/kwon.so/ec_lr_AvH.rds")

# top diff accessible peaks
top_ec_TvH <- ec_lr_TvH[ec_lr_TvH$p_val_adj < 0.05 & ec_lr_TvH$avg_log2FC > 1.5, ]
top_ec_AvH <- ec_lr_AvH[ec_lr_AvH$p_val_adj < 0.05 & ec_lr_AvH$avg_log2FC > 1, ]

# enriched motifs
ec_TvH_motif <- FindMotifs(ec_seu,
                          features = rownames(top_ec_TvH))

ec_AvH_motif <- FindMotifs(ec_seu,
                          features = rownames(top_ec_AvH))

# filter for significant motifs

```

### EC pathway enrichment 
```{r}
open_ec_TvH <- ClosestFeature(ec_seu, rownames(top_ec_TvH))
open_ec_AvH <- ClosestFeature(ec_seu, rownames(top_ec_AvH))

ec_TvH_ego <- enrichGO(gene = open_ec_TvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(ec_TvH_ego,showCategory = 20)


ec_AvH_ego <- enrichGO(gene = open_ec_AvH$gene_name,
                        keyType = "SYMBOL",
                        OrgDb = org.Mm.eg.db,
                        ont = "BP",
                        pAdjustMethod = "BH",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE)

barplot(ec_AvH_ego, showCategory = 20)
```




```{r eval=FALSE, include=FALSE}
# save csv of significant enriched motifs
# write_csv(sig_cm_4T_4H, file = "./motif_results/CM_4Tv4H.csv")
# write_csv(sig_cm_28M_4H, file = "./motif_results/CM_28Mv4H.csv")
# 
# write_csv(sig_fb_4T, file = "./motif_results/FB_4H_4T.csv")
# write_csv(sig_fb_28M, file = "./motif_results/FB_4H_28M.csv")
# 
# write_csv(sig_ec_4Hvs4T, file = "./motif_results/EC_4H_4T.csv")
# write_csv(sig_ec_4Hvs28M, file = "./motif_results/EC_4H_28M.csv")
```


